File: backends/ruby-rails/app/controllers/meta_controller.rb
class MetaController < ApplicationController
  skip_before_action :authenticate_optional

  def healthz
    head :ok
  end

  def readyz
    started_at = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    Stackverse::Sql.connection.execute("select 1")
    log_readiness(true)
    head :ok
  rescue StandardError => e
    duration_ms = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - started_at) * 1000).round
    if log_readiness(false)
      Stackverse::EventLog.warn(
        "dependency_call_failed",
        "failure",
        "Readiness lost: database unreachable",
        dependency: "postgres",
        duration_ms: duration_ms,
        error_code: e.class.name.demodulize.underscore
      )
    end
    head :service_unavailable
  end

  private

  def log_readiness(ready)
    previous = self.class.ready
    return false if previous == ready

    self.class.ready = ready
    Rails.logger.info("Readiness restored: database reachable again") if ready
    true
  end

  class << self
    attr_accessor :ready
  end
  self.ready = true
end

File: backends/php-laravel/app/Http/Controllers/MetaController.php
<?php

namespace App\Http\Controllers;

use App\Auth\Caller;
use App\Support\Logger;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;
use Throwable;

class MetaController extends Controller
{
    private static bool $wasReady = true;

    public function me(Request $request): array
    {
        return Caller::require($request)->toMeResponse();
    }

    public function healthz(): array
    {
        return ['status' => 'up'];
    }

    public function readyz(): JsonResponse
    {
        $started = microtime(true);
        try {
            DB::select('select 1');
            if (! self::$wasReady) {
                self::$wasReady = true;
                Logger::event('info', 'dependency_recovered', 'success', 'Readiness restored: database reachable again', [
                    'dependency' => 'postgres',
                ]);
            }

            return response()->json(['status' => 'ready']);
        } catch (Throwable $error) {
            if (self::$wasReady) {
                self::$wasReady = false;
                Logger::event('warn', 'dependency_call_failed', 'failure', 'Readiness lost: database unreachable', [
                    'dependency' => 'postgres',
                    'duration_ms' => (int) round((microtime(true) - $started) * 1000),
                    'error_code' => 'connection_error',
                ]);
            }

            return response()->json(['status' => 'unavailable'], 503);
        }
    }
}
